src/server/wesnothd/server.cpp

```diff
@@ -1742,7 +1742,22 @@ void server::handle_player_in_game(player_iterator p, simple_wml::document& data
 						source = "Default";
 					}
 				}
-				user_handler_->db_insert_game_player_info(uuid_, g.db_id(), side["player_id"].to_string(), side["side"].to_int(), side["is_host"].to_bool(), side["faction"].to_string(), version, source, side["current_player"].to_string());
+
+				// add leader data
+				std::vector<std::string> leaders;
+				if(side.attr("type") != "") {
+					leaders.emplace_back(side.attr("type").to_string());
+				}
+				for(const auto unit : side.children("unit")) {
+					if(unit->attr("canrecruit") == "yes") {
+						leaders.emplace_back(unit->attr("type").to_string());
+					}
+				}
+				for(const auto leader : side.children("leader")) {
+					leaders.emplace_back(leader->attr("type").to_string());
+				}
```

Comment:
What do you mean? It's checking the [side] from the scenario config.

Reply:
The client has logic to decide what leader units there are, correct? This code here duplicates that logic. Whenever the client logic changes this duplicate needs to be updated or it will be wrong. Thus it would be ideal if the logic would be in a function so that client and server can simply use the same code. Perhaps using the exact same code is not possible so then the next best thing is to add a comment that says this code here needs to stay in sync with the client side code at wherever.

Reply:
I'm still not following. The client has logic to determine which single leader to display in the MP staging screen, but I don't know why that would be relevant here.

Reply:
This is reimplementing client side logic, correct?

Reply:
Not to my knowledge.

Reply:
So the logic implemented here is completely made up? It has nothing to do with how the client decides leader units? That makes no sense at all then.

Reply:
What do you mean by "how the client decides leader units"? I'm confused why the client is relevant here in the first place. Sure, the client allows players to chose a singular leader unit (if the scenario allows it), but that's it - if there are multiple leader units defined then that's only in the [side] definition.

Reply:
With the client I mean the wesnoth program. The program that among other things reads the scenario config, decides what it all means. For example what leader units to create etc. Now you want to replicate part of that logic to also decide what leader units exist in a game in wesnothd. Does that make any sense so far?

Reply:
I assume you're referring to the logic in flg_manager's constructor? The logic added to wesnothd is certainly not a replication of that.

Reply:
I do understand the idea of having a consistent way of detecting what leaders there are, but as far as I can tell the client simply doesn't check that in single place in the code in a way that matches what would need be checked in wesnothd. If you know where this check is replicating code that already exists on the client, then please let me know where it is so I can take a look at making it common code with wesnothd.

Reply:
I have no idea where the client side logic is but surely you must have written this logic based on something. If it's not a single place on the client side that certainly makes it more annoying but then those places should be documented so there is a better chance to keep this code uptodate.

Reply:
I wrote this based on the wiki and asking in the #development channel for confirmation, and then tested it locally.

To (mostly) quote myself from that discussion:
* `[side]type=` works as you'd expect it to - you see that unit in MP staging and get it as your single leader.
* `[side][leader]` without `no_leader=yes` gives you two leaders - the leader chosen in MP staging as well as the one specified by [leader].
* `[side][leader]` with `no_leader=yes` gives you the leader from [leader], but you don't see that on MP staging and while it lets you choose a faction the leader selection dropdown is disabled.
* `[side][unit]` with canrecruit=yes works as you'd expect it to - you see that unit in MP staging and get it as your single leader. however if you don't specify an x,y or `placement=leader` then you immediately lose because your leader failed to spawn.

So a big part of the problem is also that the client itself is not consistent with how it handles or displays these different cases. I did not check what it does when you start combining multiple ways to set a leader unit in a single [side].

Reply:
Well, it'd be nice to add any kind of hint to understand the intention of this code. As I understand it the goal is to figure out what `canrecruit=yes` units the client creates at the start of the game? At least that is my understanding of what leader units are. (And of course we can only record the initial state.)

What the client shows before the game starts as the (single) leader could also be the goal if we wanted to just store one leader.

Reply:
Since it doesn't look like we can share code or even point to code in the client that does the same thing, yes.